### ISSUE TITLE
New interface for arbitrary group operations

### ISSUE THREAD

**Issue by fhenslee** [static-frame/static-frame#695]
**Description**
Requesting the addition of a new series group iter, which is referred to here as "iter_arbitrary_group" (but can be renamed to fit into sf's terminology). Perhaps this can be boiled up to the frame level too.

The interface can take an iterable with len = series.shape[0] and group according to its values. This allows for the creation of groups outside of those defined by a series' index or values.

**Example**
```python
import numpy as np
import static_frame as sf

index = sf.IndexDate.from_year_range(
    start=2000,
    stop=2001,
)

series = sf.Series(
    values=np.ones(shape=index.shape[0]),
    index=index,
    name='ones',
)

# proposed interface
annual_sum = series.iter_arbitrary_group(iterable=series.index.via_dt.year).apply(np.sum)

# how to do it currently
fgo = series.to_frame_go()
fgo['year'] = fgo.index.via_dt.year
expected_output = fgo.iter_group(key='year').apply(lambda f: f['ones'].sum())
```

**Platform**
```python
>>> import static_frame as sf
>>> sf.Platform.display()

platform           Windows-10-10.0.19041-SP0
sys.version        3.8.3 | packaged by conda-forge | (default, Jun  1 2020, 16:59:10) [MSC v.1916 64 bit (AMD64)]
static-frame       0.9.19
numpy              1.19.5
pandas             0.25.3
xlsxwriter         1.1.2
openpyxl           3.0.9
xarray             <ModuleNotFoundError>
tables             3.6.1
pyarrow            0.17.0
msgpack            <ModuleNotFoundError>
msgpack_numpy      <ModuleNotFoundError>
```

**Comment by flexatone** [comment]
Many thank this for this suggestion; this fills a gap in functionality I have been aware of for some time. I will explore an implementation.

**Comment by flexatone** [comment]
This feature is now available in 0.9.23 as `iter_group_other()` and related interfaces:

```python
>>> index = sf.IndexDate.from_year_range(2000, 2001)
>>> s = sf.Series(np.arange(len(index)), index=index)
>>> s.iter_group_other(s.index.via_dt.year).apply(np.sum)
<Series>
<Index>
2000     66795
2001     200020
<int64>  <int64>
```
